[WFM] Issue in message deserialisation

At least one set of messages lead to unhandled error into "CTRL" code.

```
2017-12-13 12:24:41.787 o.o.w.t.s.StatsTopology [ERROR] Unhandled exception
org.openkilda.wfm.MessageFormatException: Invalid input message/tuple
	at org.openkilda.wfm.protocol.JsonMessage.<init>(JsonMessage.java:28) ~[stormjar.jar:?]
	at org.openkilda.wfm.protocol.KafkaMessage.<init>(KafkaMessage.java:12) ~[stormjar.jar:?]
	at org.openkilda.wfm.topology.event.PopulateIslFilterAction.handle(PopulateIslFilterAction.java:30) ~[stormjar.jar:?]
	at org.openkilda.wfm.AbstractAction.run(AbstractAction.java:25) [stormjar.jar:?]
	at org.openkilda.wfm.topology.event.OFELinkBolt.doWork(OFELinkBolt.java:138) [stormjar.jar:?]
	at org.openkilda.wfm.topology.utils.AbstractTickStatefulBolt.execute(AbstractTickStatefulBolt.java:84) [stormjar.jar:?]
	at org.apache.storm.topology.StatefulBoltExecutor.doExecute(StatefulBoltExecutor.java:145) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.topology.StatefulBoltExecutor.handleTuple(StatefulBoltExecutor.java:137) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.topology.BaseStatefulBoltExecutor.execute(BaseStatefulBoltExecutor.java:78) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.daemon.executor$fn__9866$tuple_action_fn__9868.invoke(executor.clj:730) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.daemon.executor$mk_task_receiver$fn__9787.invoke(executor.clj:462) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.disruptor$clojure_handler$reify__9301.onEvent(disruptor.clj:40) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:472) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.utils.DisruptorQueue.consumeBatchWhenAvailable(DisruptorQueue.java:451) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.disruptor$consume_batch_when_available.invoke(disruptor.clj:73) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.daemon.executor$fn__9866$fn__9879$fn__9932.invoke(executor.clj:849) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at org.apache.storm.util$async_loop$fn__555.invoke(util.clj:484) [storm-core-1.1.0.2.6.2.0-205.jar:1.1.0.2.6.2.0-205]
	at clojure.lang.AFn.run(AFn.java:22) [clojure-1.7.0.jar:?]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "flowid" (class org.openkilda.messaging.model.ImmutablePair), not marked as ignorable (2 known properties: "reverse", "forward"])
 at [Source: {
    "timestamp": 1513159401394, 
    "correlation_id": "topology_reader-1513159401.39", 
    "destination": "WFM_CACHE", 
    "type": "INFO", 
    "payload": {
        "operation": "CACHE", 
        "message_type": "flow_operation", 
        "payload": {
            "forward": {
                "src_port": 1, 
                "bandwidth": 2000, 
                "flowid": "pre-populated-frw1", 
                "dst_switch": "00:00:00:00:ab:00:00:02", 
                "src_switch": "00:00:00:00:ab:00:00:01", 
                "cookie": 22213, 
                "dst_port": 2, 
                "src_vlan": 0, 
                "dst_vlan": 0
            }, 
            "flowid": "pre-populated-1", 
            "reverse": {}
        }
    }
}; line: 23, column: 10] (through reference chain: org.openkilda.messaging.info.InfoMessage["payload"]->org.openkilda.messaging.info.flow.FlowInfoData["payload"]->org.openkilda.messaging.model.ImmutablePair["flowid"])
	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:62) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnknownProperty(DeserializationContext.java:834) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:1093) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1478) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperties(BeanDeserializerBase.java:1432) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:421) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:511) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:400) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:183) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:150) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:129) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:97) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:142) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:502) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:511) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:400) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:183) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:150) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:129) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:97) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeWithType(BeanDeserializerBase.java:1082) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:63) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3798) ~[stormjar.jar:?]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2842) ~[stormjar.jar:?]
	at org.openkilda.wfm.protocol.KafkaMessage.unpackJson(KafkaMessage.java:21) ~[stormjar.jar:?]
	at org.openkilda.wfm.protocol.KafkaMessage.unpackJson(KafkaMessage.java:10) ~[stormjar.jar:?]
	at org.openkilda.wfm.protocol.JsonMessage.<init>(JsonMessage.java:26) ~[stormjar.jar:?]
	... 18 more
```

--- a/services/topology-engine/queue-engine/topologies/example-topology-prepopulation.json
+++ b/services/topology-engine/queue-engine/topologies/example-topology-prepopulation.json
@@ -39,7 +39,6 @@
     ],
     "flows": [
         {
-            "flowid": "pre-populated-1",
             "forward": {
                 "flowid": "pre-populated-frw1",
                 "cookie": 22213,

--- a/services/wfm/src/main/java/org/openkilda/wfm/isl/DiscoveryManager.java
+++ b/services/wfm/src/main/java/org/openkilda/wfm/isl/DiscoveryManager.java
@@ -103,7 +103,7 @@ public class DiscoveryManager {
             return;
         }
 
-        subject = new DiscoveryNode(node.switchId);
+        subject = new DiscoveryNode(node.switchId, node.portId);
         pollQueue.add(subject);
         logger.info("New {}", subject);
     }

--- a/services/wfm/src/main/java/org/openkilda/wfm/isl/DiscoveryNode.java
+++ b/services/wfm/src/main/java/org/openkilda/wfm/isl/DiscoveryNode.java
@@ -4,13 +4,14 @@ import java.io.Serializable;
 import java.util.Objects;
 
 public class DiscoveryNode implements Serializable {
+    // FIXME(surabujin): ... is there any better way to define this value?
     private final static int NEVER = 999999;
     private final String switchId;
     private final String portId;
     private int age;
     private int timeCounter;
     private int checkInterval;
-    private int consequitiveFailures;
+    private int consecutiveFailures;
     private int forlornThreshold;
 
 
@@ -21,17 +22,17 @@ public class DiscoveryNode implements Serializable {
         this.timeCounter = 0;
         this.checkInterval = checkInterval;
         this.forlornThreshold = forlornThreshold;
-        consequitiveFailures = 0;
+        consecutiveFailures = 0;
     }
 
     public DiscoveryNode(String switchId, String portId) {
         //FIXME: extract checkInterval & forlornThreshold from a config.
-        this(switchId, portId, 1, NEVER);
+        this(switchId, portId, 0, NEVER);
     }
 
     public DiscoveryNode(String switchId) {
         //FIXME: extract checkInterval & forlornThreshold from a config.
-        this(switchId, "", 1, NEVER);
+        this(switchId, "", 0, NEVER);
     }
 
     public void renew() {
@@ -43,11 +44,11 @@ public class DiscoveryNode implements Serializable {
         if (forlornThreshold == NEVER) { // never gonna give a link up.
              return false;
         }
-        return consequitiveFailures >= forlornThreshold;
+        return consecutiveFailures >= forlornThreshold;
     }
 
     public void redeem() {
-        consequitiveFailures = 0;
+        consecutiveFailures = 0;
     }
 
     public void incAge() {
@@ -67,7 +68,7 @@ public class DiscoveryNode implements Serializable {
     }
 
     public void countFailure() {
-        consequitiveFailures++;
+        consecutiveFailures++;
     }
 
     public boolean timeToCheck() {

--- a/services/wfm/src/main/java/org/openkilda/wfm/topology/event/PopulateIslFilterAction.java
+++ b/services/wfm/src/main/java/org/openkilda/wfm/topology/event/PopulateIslFilterAction.java
@@ -1,7 +1,5 @@
 package org.openkilda.wfm.topology.event;
 
-import com.fasterxml.jackson.core.JsonProcessingException;
-import org.apache.storm.tuple.Tuple;
 import org.openkilda.messaging.Destination;
 import org.openkilda.messaging.Message;
 import org.openkilda.messaging.command.CommandMessage;
@@ -13,6 +11,9 @@ import org.openkilda.wfm.MessageFormatException;
 import org.openkilda.wfm.UnsupportedActionException;
 import org.openkilda.wfm.isl.DummyIIslFilter;
 import org.openkilda.wfm.protocol.KafkaMessage;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.apache.storm.tuple.Tuple;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -50,5 +51,24 @@ public class PopulateIslFilterAction extends AbstractAction {
             logger.info("Add ISL filter record - switcID=\"{}\" portId=\"{}\"", entity.switchId, entity.portId);
             filter.add(entity.switchId, entity.portId);
         }
+    }
+
+    @Override
+    protected Boolean handleError(Exception e) {
+        boolean isHandled = true;
+
+        try {
+            throw e;
+        } catch (MessageFormatException exc) {
+            logger.error("Can\'t unpack input tuple: {}", exc.getCause().getMessage());
+
+            for (int i = 0; i < getTuple().size(); i++) {
+                logger.error("Field #{}: {}", i, getTuple().getValue(i));
+            }
+        } catch (Exception exc) {
+            isHandled = false;
+        }
+
+        return isHandled;
     }
 }

--- a/services/wfm/src/test/java/org/openkilda/wfm/topology/event/OFELinkBoltTest.java
+++ b/services/wfm/src/test/java/org/openkilda/wfm/topology/event/OFELinkBoltTest.java
@@ -1,0 +1,48 @@
+package org.openkilda.wfm.topology.event;
+
+import org.openkilda.wfm.AbstractStormTest;
+import org.openkilda.wfm.ConfigurationException;
+import org.openkilda.wfm.protocol.KafkaMessage;
+import org.openkilda.wfm.topology.OutputCollectorMock;
+import org.openkilda.wfm.topology.TopologyConfig;
+
+import org.apache.storm.state.InMemoryKeyValueState;
+import org.apache.storm.task.OutputCollector;
+import org.apache.storm.task.TopologyContext;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.TupleImpl;
+import org.apache.storm.tuple.Values;
+import org.junit.Test;
+import org.kohsuke.args4j.CmdLineException;
+import org.mockito.Mockito;
+
+public class OFELinkBoltTest extends AbstractStormTest {
+    private static final Integer TASK_ID_BOLT = 0;
+    private static final String COMPONENT_ID_SOURCE = OFEventWFMTopology.SPOUT_ID_INPUT;
+    private static final String STREAM_ID_INPUT = "input";
+
+    @Test
+    public void invalidJsonForDiscoveryFilter() throws CmdLineException, ConfigurationException {
+        OFEventWFMTopology manager = new OFEventWFMTopology(makeLaunchEnvironment());
+        TopologyConfig config = manager.getConfig();
+        OFELinkBolt bolt = new OFELinkBolt(config);
+
+        TopologyContext context = Mockito.mock(TopologyContext.class);
+
+        Mockito.when(context.getComponentId(TASK_ID_BOLT))
+                .thenReturn(COMPONENT_ID_SOURCE);
+        Mockito.when(context.getComponentOutputFields(COMPONENT_ID_SOURCE, STREAM_ID_INPUT))
+                .thenReturn(KafkaMessage.FORMAT);
+
+        OutputCollectorMock outputDelegate = Mockito.spy(new OutputCollectorMock());
+        OutputCollector output = new OutputCollector(outputDelegate);
+
+        bolt.prepare(stormConfig(), context, output);
+        bolt.initState(new InMemoryKeyValueState<>());
+
+        Tuple tuple = new TupleImpl(context, new Values("{\"corrupted-json"), TASK_ID_BOLT, STREAM_ID_INPUT);
+        bolt.doWork(tuple);
+
+        Mockito.verify(outputDelegate).ack(tuple);
+    }
+}

